pkg/route/controller/ingress/ingress.go

```diff
@@ -0,0 +1,591 @@
+//    that exists, a route should exist with a valid TLS config from that secret.
+// 3. For every service referenced by the ingress path rule, the route should have
+//    an update to date target port based on the service.
+// 4. A route owned by an ingress that no longer satisfies the first three invariants
+//    should be deleted.
+//
+// Unsupported attributes:
+//
+// * the ingress class attribute
+// * nginx annotations
+// * the empty backend
+// * paths with empty backends
+// * creating a dynamic route spec.host
+//
+type Controller struct {
+	eventRecorder record.EventRecorder
+
+	// Allows injection for testing, controls requeues on errors
+	resourceFailureDelayFn func(requeue int) (time.Duration, bool)
+
+	client        routeclient.RoutesGetter
+	secretLister  corelisters.SecretLister
+	ingressLister extensionslisters.IngressLister
+	routeLister   routelisters.RouteLister
+	serviceLister corelisters.ServiceLister
+
+	// queue is the list of namespace keys that must be synced.
+	queue workqueue.RateLimitingInterface
+
+	// syncs are the items that must return true before the queue can be processed
+	syncs []cache.InformerSynced
+}
+
+type queueKey struct {
+	namespace string
+	name      string
+}
+
+// NewController instantiates a Controller
+func NewController(eventsClient kv1core.EventsGetter, client routeclient.RoutesGetter, ingresses extensionsinformers.IngressInformer, secrets coreinformers.SecretInformer, services coreinformers.ServiceInformer, routes routeinformers.RouteInformer) *Controller {
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartLogging(glog.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	broadcaster.StartRecordingToSink(&kv1core.EventSinkImpl{Interface: eventsClient.Events("")})
+	recorder := broadcaster.NewRecorder(legacyscheme.Scheme, v1.EventSource{Component: "ingress-to-route-controller"})
+
+	c := &Controller{
+		eventRecorder: recorder,
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ingress-to-route"),
+
+		client:        client,
+		ingressLister: ingresses.Lister(),
+		secretLister:  secrets.Lister(),
+		routeLister:   routes.Lister(),
+		serviceLister: services.Lister(),
+
+		resourceFailureDelayFn: defaultResourceFailureDelay,
+	}
+
+	processNamespaceFn := func(obj interface{}) {
+		switch t := obj.(type) {
+		case metav1.Object:
+			ns := t.GetNamespace()
+			if len(ns) == 0 {
+				utilruntime.HandleError(fmt.Errorf("object %T has no namespace", obj))
+				return
+			}
+			c.queue.Add(queueKey{namespace: ns})
+		default:
+			utilruntime.HandleError(fmt.Errorf("couldn't get key for object %T", obj))
+		}
+	}
+
+	// any change to a secret of type TLS in the namespace
+	secrets.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			switch t := obj.(type) {
+			case *v1.Secret:
+				return t.Type == v1.SecretTypeTLS
+			}
+			return true
```

Comment:
if you're not a secret, why bother queuing.  It won't work right.

Reply:
tombstones?